runtime/queries/metricsview_comparison_toplist.go

```diff
@@ -422,16 +428,37 @@ func (q *MetricsViewComparisonToplist) buildMetricsComparisonTopListSQL(mv *runt
 		twiceTheLimitClause = fmt.Sprintf(" LIMIT %d", q.Limit*2)
 	}
 
+	baseLimitClause := ""
+	comparisonLimitClause := ""
+
+	joinType := "FULL"
+	if !q.Exact {
+		approximationLimit := q.Limit
+		if q.Limit < 100 {
+			approximationLimit = 100
+		}
+
+		if q.Sort[0].Type == runtimev1.MetricsViewComparisonSortType_METRICS_VIEW_COMPARISON_SORT_TYPE_BASE_VALUE ||
+			q.Sort[0].Type == runtimev1.MetricsViewComparisonSortType_METRICS_VIEW_COMPARISON_SORT_TYPE_ABS_DELTA ||
+			q.Sort[0].Type == runtimev1.MetricsViewComparisonSortType_METRICS_VIEW_COMPARISON_SORT_TYPE_REL_DELTA {
+			joinType = "LEFT OUTER"
+			baseLimitClause = fmt.Sprintf("LIMIT %d", approximationLimit)
+		} else if q.Sort[0].Type == runtimev1.MetricsViewComparisonSortType_METRICS_VIEW_COMPARISON_SORT_TYPE_COMPARISON_VALUE {
+			joinType = "RIGHT OUTER"
+			comparisonLimitClause = fmt.Sprintf("LIMIT %d", approximationLimit)
+		}
+	}
+
 	var sql string
 	if dialect != drivers.DialectDruid {
 		sql = fmt.Sprintf(`
 		SELECT COALESCE(base.%[2]s, comparison.%[2]s) AS %[10]s, %[9]s FROM 
 			(
-				SELECT %[1]s FROM %[3]q WHERE %[4]s GROUP BY %[2]s
+				SELECT %[1]s FROM %[3]q WHERE %[4]s GROUP BY %[2]s %[12]s 
 			) base
-		FULL JOIN
+		%[11]s JOIN
 			(
-				SELECT %[1]s FROM %[3]q WHERE %[5]s GROUP BY %[2]s
+				SELECT %[1]s FROM %[3]q WHERE %[5]s GROUP BY %[2]s %[13]s 
 			) comparison
```

Comment:
Not sure I follow this – it seems like it's applying the approximation limit *before* doing the ordering?

Reply:
good catch